sql-plugin/src/main/java/com/nvidia/spark/rapids/GpuColumnVector.java

```diff
@@ -280,29 +281,30 @@ public RapidsHostColumnBuilder builder(int i) {
     @Override
     protected ai.rapids.cudf.ColumnVector buildAndPutOnDevice(int builderIndex) {
       ai.rapids.cudf.ColumnVector cv = builders[builderIndex].buildAndPutOnDevice();
+      builders[builderIndex].close();
       builders[builderIndex] = null;
       return cv;
     }
 
     public HostColumnVector[] buildHostColumns() {
-      HostColumnVector[] vectors = new HostColumnVector[builders.length];
-      try {
-        for (int i = 0; i < builders.length; i++) {
-          vectors[i] = builders[i].build();
+      // buildHostColumns is called from tryBuild, and tryBuild has to be safe to call
+      // multiple times, so if a retry exception happens in this code, we need to pick
+      // up where we left off last time.
+      if (wipHostColumns == null) {
+        wipHostColumns = new HostColumnVector[builders.length];
+      }
+      for (int i = 0; i < builders.length; i++) {
+        if (builders[i] != null && wipHostColumns[i] == null) {
+          wipHostColumns[i] = builders[i].build();
+          builders[i].close();
           builders[i] = null;
-        }
-        HostColumnVector[] result = vectors;
-        vectors = null;
-        return result;
-      } finally {
-        if (vectors != null) {
-          for (HostColumnVector v : vectors) {
-            if (v != null) {
-              v.close();
-            }
-          }
+        } else if (builders[i] == null && wipHostColumns[i] == null) {
+          throw new IllegalStateException("buildHostColumns cannot be called more than once");
```

Comment:
this exception message seems to go against the comment at the top of this function (`buildHostColumns`), but I can see why you need to throw here. Maybe "`buildHostColumns` is in an invalid state due to a retry"?

Reply:
So here is where things are odd. The old code had some APIs that could be called multiple times and other APIs that could not, but in reality none of them could be called multiple times. I fixed try build, but this is called from try build. try build caches the result of calling this, if it succeeded.  So this odd function needs to be able to be called again if an exception was thrown while it was running. But if it succeeded, then it cannot be called again, partly because the builders themselves follow this same behavior.

I can fix all of them to be able to be called as many times as we want and even add more rows in between building new HostColumnVectors.  The problem is that is a much larger change and does not provide any real value at this time.

Reply:
The fact that you have this comment here on github is good enough for me.